distributed/shuffle/tests/test_shuffle.py

```diff
@@ -469,3 +476,208 @@ async def test_clean_after_close(c, s, a, b):
+        return out
+
+    def new_shuffle(
+        self, name, worker_for_mapping, schema, directory, loop, Shuffle=Shuffle
+    ):
+        s = Shuffle(
+            column="_partition",
+            worker_for=worker_for_mapping,
+            # FIXME: Is output_workers redundant with worker_for?
+            output_workers=set(worker_for_mapping.values()),
+            schema=schema,
+            directory=directory / name,
+            id=ShuffleId(name),
+            local_address=name,
+            nthreads=2,
+            rpc=self,
+            broadcast=self.fake_broadcast,
+            memory_limiter_disk=ResourceLimiter(10000000),
+            memory_limiter_comms=ResourceLimiter(10000000),
+        )
+        self.shuffles[name] = s
+        return s
+
+
+# 36 parametrizations
+# Runtime each ~0.1s
+@pytest.mark.parametrize("n_workers", [1, 10])
+@pytest.mark.parametrize("n_input_partitions", [1, 2, 10])
+@pytest.mark.parametrize("npartitions", [1, 20])
+@pytest.mark.parametrize("barrier_first_worker", [True, False])
+@gen_test(timeout=5)
+async def test_basic_lowlevel_shuffle(
+    tmpdir,
+    loop_in_thread,
+    n_workers,
+    n_input_partitions,
+    npartitions,
+    barrier_first_worker,
+):
+    dfs = []
+    rows_per_df = 10
+    for ix in range(n_input_partitions):
+        df = pd.DataFrame({"x": range(rows_per_df * ix, rows_per_df * (ix + 1))})
+        df["_partition"] = df.x % npartitions
+        dfs.append(df)
+
+    workers = list("abcdefghijklmn")[:n_workers]
+
+    worker_for_mapping = {}
+
+    for part in range(npartitions):
+        worker_for_mapping[part] = get_worker_for(part, workers, npartitions)
+    assert len(set(worker_for_mapping.values())) == min(n_workers, npartitions)
+    schema = pa.Schema.from_pandas(dfs[0])
+
+    local_shuffle_pool = ShuffleTestPool()
+    shuffles = []
+    for ix in range(n_workers):
+        shuffles.append(
+            local_shuffle_pool.new_shuffle(
+                name=workers[ix],
+                worker_for_mapping=worker_for_mapping,
+                schema=schema,
+                directory=tmpdir,
+                loop=loop_in_thread,
+            )
+        )
+    random.seed(42)
+    if barrier_first_worker:
+        barrier_worker = shuffles[0]
+    else:
+        barrier_worker = random.sample(shuffles, k=1)[0]
+
+    try:
+        for ix, df in enumerate(dfs):
+            s = shuffles[ix % len(shuffles)]
+            await s.add_partition(df)
+
+        await barrier_worker.barrier()
+
+        total_bytes_sent = 0
+        total_bytes_recvd = 0
+        total_bytes_recvd_shuffle = 0
+        for s in shuffles:
+            metrics = s.heartbeat()
+            assert metrics["comm"]["total"] == metrics["comm"]["written"]
+            total_bytes_sent += metrics["comm"]["written"]
+            total_bytes_recvd += metrics["disk"]["total"]
+            total_bytes_recvd_shuffle += s.total_recvd
+
+        assert total_bytes_recvd_shuffle == total_bytes_sent
```

Comment:
```suggestion
        assert total_bytes_recvd_shuffle == total_bytes_sent == total_bytes_recvd
```

From what I understand, we probably want to test that as well.

Reply:
these are not identical. there appears to be a small drift between sizes when comparing pyarrow buffers and bytes directly. Therefore, what the comms measure is slightly different from what is actually received. This is very ugly but I'm not willing to fix this right now. Refactoring the serialization parts is out of scope

Reply:
Let's highlight this in the test then or drop `total_bytes_recvd` altogether. Right now it looks like we just forgot to do something useful with `total_bytes_recvd`.